src/core/handle.ml

```diff
@@ -258,50 +258,103 @@ let handle_cmd : sig_state -> p_command -> sig_state * proof_data option =
       out 3 "(symb) %s ≔ %a\n" x.elt pp_term t;
       let d = if op then None else Some(t) in
       (fst (Sig_state.add_symbol ss expo Defin Eager x a impl d), None)
-  | P_inductive(ms, id, a, l)     ->
-      (* Verify the modifiers. *)
+  | P_inductive(ms, il)     ->
+      (* STEP 0 - Verify the modifiers. *)
       let (prop, e, mstrat) = handle_modifiers ms in
       if prop <> Defin then
         fatal cmd.pos "Property modifiers cannot be used in inductive types.";
       if mstrat <> Eager then
         fatal cmd.pos "Pattern matching strategy modifiers cannot be used \
                        in inductive types.";
-      (* Add the inductive type in the signature *)
-      let (ss, ind_sym) = handle_symbol ss e Injec Eager id [] a in
-      (* Add the constructors in the signature.  *)
-      let add_cons :
-            sig_state * sym list -> ident * p_term -> sig_state * sym list
-        = fun (ss, cons_list) (id, a) ->
-        let (ss, cons_sym) = handle_symbol ss e Const Eager id [] a in
-        (ss, cons_sym::cons_list)
+      (* STEP 1 - Add the inductive types in the signature *)
+      let add_ind_types :
+        sig_state * sym list -> (ident * p_term * (ident * p_term) list) loc
+        -> sig_state * sym list
+        = fun (ss, ind_typ_list) x ->
+        let (id, a, _) = x.elt in
+        let (ss, ind_sym) = handle_symbol ss e Injec Eager id [] a in
+        (ss, ind_sym::ind_typ_list)
+      in
+      let (ss, ind_typ_list_rev) = List.fold_left add_ind_types (ss, []) il in
+      let ind_typ_list = List.rev ind_typ_list_rev in
+      (* STEP 2 - Add the constructors in the signature. *)
+      let add_constructors :
+        sig_state*(sym list) list -> (ident*p_term*(ident*p_term) list) loc
+        -> sig_state*(sym list) list
+        = fun (ss, cons_list_list) x ->
+        let (_, _, l) = x.elt in
+        let add_cons :
+              sig_state * sym list -> ident * p_term -> sig_state * sym list
+          = fun (ss, cons_list) (id, a) ->
+          let (ss, cons_sym) = handle_symbol ss e Const Eager id [] a in
+          (ss, cons_sym::cons_list)
+        in
+        let (ss, cons_list_rev) = List.fold_left add_cons (ss, []) l in
+        (* Reverse the list of constructors previously computed to preserve
+           the initial order *)
+        let cons_list = List.rev cons_list_rev in
+        (ss, cons_list::cons_list_list)
+      in
+      let (ss, cons_list_list_rev) =
+        List.fold_left add_constructors (ss, []) il
+      in
+      let cons_list_list = List.rev cons_list_list_rev in
+      (* STEP 3 - Generate the induction principle. *)
+        (* A - Compute the induction principle *)
+      let rec_typ_list, assoc_predicats =
+        Inductive.gen_rec_type ss cmd.pos ind_typ_list cons_list_list
```

Comment:
It would be better that gen_rec_type takes as argument `(sym * sym list) list`.

Reply:
Why it's better?
Do you really want to go through our list again just to avoid recursive calls to multiple lists that are not the same length? They are by construction. I don't understand why sometimes you have to optimize, sometimes not, etc.

Reply:
Less invariants to take care.

Reply:
Another reason is that the structure output by the parser is p_inductive list with p_inductive = ident * p_term * (ident * p_term) list: each inductive type comes with its constructors. It would be nice to keep a similar structure everywhere.

Reply:
Done! :)